source/common/config/grpc_mux_impl.h

```diff
@@ -37,12 +36,18 @@ class GrpcMuxImpl
   void pause(const std::string& type_url) override;
   void resume(const std::string& type_url) override;
 
-  void sendDiscoveryRequest(const std::string& type_url) override;
+  void sendDiscoveryRequest(const std::string& type_url);
 
-  // GrpcStream
-  void handleResponse(std::unique_ptr<envoy::api::v2::DiscoveryResponse>&& message) override;
+  // Config::XdsGrpcContext
   void handleStreamEstablished() override;
   void handleEstablishmentFailure() override;
+  void drainRequests() override;
+
+  void handleDiscoveryResponse(std::unique_ptr<envoy::api::v2::DiscoveryResponse>&& message);
+  GrpcStream<envoy::api::v2::DiscoveryRequest, envoy::api::v2::DiscoveryResponse>&
+  grpcStreamTestOnly() {
```

Comment:
I think this can be avoided by having another constructor that takes the `grpc_stream_` as an arg, i.e. doing dependency injection. You can then have the real constructor call this one with the real `grpc_stream_`.

Reply:
The problem with that is that GrpcStream needs a reference to GrpcStreamCallbacks (was XdsGrpcContext).... which is the GrpcMux, which has to make the GrpcStream within its constructor. Separating out like this would require some sort of set_grpc_stream() for GrpcMux... although maybe GrpcMuxImpl::start() and GrpcStream::establishNewStream() could fill that role? I'll look into that if that sounds sensible to you.

Reply:
Yeah, I was thinking more along the lines of injecting a mock `GrpcStream`, although looking at how it is used in tests, maybe you actually want the real thing.

One last nit: you rename this to `grpcStreamForTest`?

Reply:
Done.